internal/worker/worker.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		hostname = ""
+	}
+
+	const (
+		ReservedLabelName         = "name"
+		ReservedLabelVersion      = "version"
+		ReservedLabelHostname     = "hostname"
+		ReservedLabelOS           = "os"
+		ReservedLabelArchitecture = "arch"
+	)
+
+	// Create base labels
+	labels := map[string]string{
+		ReservedLabelName:         worker.name,
+		ReservedLabelVersion:      version.FullVersion,
+		ReservedLabelHostname:     hostname,
+		ReservedLabelOS:           runtime.GOOS,
+		ReservedLabelArchitecture: runtime.GOARCH,
+	}
+
+	// Merge with the user specified labels
+	for key, value := range worker.userSpecifiedLabels {
+		if _, alreadyExists := labels[key]; !alreadyExists {
+			labels[key] = value
+		}
+	}
+
+	return &api.WorkerInfo{
+		Labels: labels,
+	}
+}
+
+func (worker *Worker) Run(ctx context.Context) error {
+	// A sub-context to cancel out all Run() side-effects when it finishes
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var rpcSecurity grpc.DialOption
+
+	if worker.rpcInsecure {
+		rpcSecurity = grpc.WithInsecure()
+	} else {
+		certPool, _ := gocertifi.CACerts()
+		tlsCredentials := credentials.NewTLS(&tls.Config{
+			MinVersion: tls.VersionTLS13,
+			RootCAs:    certPool,
+		})
+		rpcSecurity = grpc.WithTransportCredentials(tlsCredentials)
+	}
+
+	// https://github.com/grpc/grpc-go/blob/master/Documentation/concurrency.md
+	conn, err := grpc.DialContext(subCtx, worker.rpcEndpoint, rpcSecurity)
+	if err != nil {
+		worker.logger.Errorf("failed to dial %s: %v", worker.rpcEndpoint, err)
+	}
+	worker.rpcClient = api.NewCirrusWorkersServiceClient(conn)
+	defer conn.Close()
+
+	for {
+		if worker.sessionToken == "" {
+			if err := worker.register(subCtx); err != nil {
+				worker.logger.Errorf("failed to register worker: %v", err)
+			}
+		} else {
+			err := worker.poll(subCtx)
+
+			if errors.Is(err, ErrShutdown) {
+				return nil
+			}
+
+			if err != nil {
+				worker.logger.Errorf("failed to poll: %v", err)
+			}
+		}
+
+		select {
+		case <-subCtx.Done():
+			return nil
+		case <-time.After(time.Duration(worker.pollIntervalSeconds) * time.Second):
+			// continue the loop
+		}
+	}
+}
+
+func (worker *Worker) register(ctx context.Context) error {
+	response, err := worker.rpcClient.Register(ctx, &api.RegisterRequest{
+		WorkerInfo:        worker.info(),
+		RegistrationToken: worker.registrationToken,
+	})
+	if err != nil {
+		return err
+	}
+
+	worker.sessionToken = response.SessionToken
+
+	worker.logger.Infof("worker successfully registered")
+
+	return nil
+}
+
+func (worker *Worker) poll(ctx context.Context) error {
+	// De-register completed tasks
+	worker.registerTaskCompletions()
+
+	worker.logger.Infof("polling %s", worker.rpcEndpoint)
+
+	request := &api.PollRequest{
+		WorkerInfo:   worker.info(),
+		RunningTasks: worker.runningTasks(),
+	}
+
+	response, err := worker.rpcClient.Poll(ctx, request, grpc.PerRPCCredentials(worker))
+	if err != nil {
+		return err
+	}
+
+	if response.Shutdown {
```

Comment:
Let's move it right after `TasksToStop` so it's possible to stop and shutdown.

Reply:
I don't think the order should matter here since all started tasks are tied to the `subCtx`, which will be closed once `Run()` exits.